Filter logging using regular expressions

petern48 opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allows users to specify regular expressions through new `message_regexp` and `message_regexp_negative` fields in the `config.xml` file to filter out logging. The logging is applied to the formatted un-colored text for the most intuitive developer experience.

These regexps can be specified globally and/or per-level, with per-level regexps overriding the global ones. See the new doc updates for an example. The filters update dynamically as the `config.xml` updates while the program runs.

Closes #69447

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

petern48 committed:
Add OwnFilteringChannel[.h, .cpp]

petern48 committed:
Implement filtering channel on top of formatting channel in Loggers.cpp

petern48 committed:
Add filter channel to LocalServer.cpp and BaseDaemon.cpp

petern48 committed:
Update docs for regexp logging feature in settings.md

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #69657 from petern48/ClickHouse:regexp_logger -->
*This is an automated comment for commit 48c1f0fd57f65fc4a0114603147ab35492e27417 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69657/48c1f0fd57f65fc4a0114603147ab35492e27417/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69657/48c1f0fd57f65fc4a0114603147ab35492e27417/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

petern48 committed:
Fix style

petern48 committed:
Apply regexp filtering to root logger

max-vostrikov commented:
Would be nice if there were an integration test added for this feature. It seems that [this integration test](https://github.com/ClickHouse/ClickHouse/blob/cf245538b91923a97c617a997e2bc68d01b51d75/tests/integration/test_log_levels_update) is testing somewhat similar.

antaljanosbenjamin commented on src/Loggers/OwnFilteringChannel.h:
```diff
@@ -0,0 +1,61 @@
+            pChannel->close();
+    }
+
+    void setProperty(const std::string& name, const std::string& value) override
+    {
+        if (pChannel)
+            pChannel->setProperty(name, value);
+    }
```
Maybe also implement `getProperty`.

petern48 replied:
Thanks for the catch. That fixed the bugs from the previous tests

antaljanosbenjamin commented on src/Loggers/Loggers.cpp:
```diff
@@ -373,6 +411,14 @@ void Loggers::updateLevels(Poco::Util::AbstractConfiguration & config, Poco::Log
                     const std::string level(config.getString("logger.levels." + key + ".level"));
                     logger.root().get(name).setLevel(level);
+
+                    std::string pos_pattern = config.getRawString("logger.levels." + key + "message_regexp", global_pos_pattern);
+                    std::string neg_pattern = config.getRawString("logger.levels." + key + "message_regexp_negative", global_neg_pattern);
```
What is expected behavior when I specify a global positive pattern and a level specific negativ pattern?

petern48 replied:
Both the global positive pattern and level-specific negative pattern should apply for the expression

antaljanosbenjamin commented on src/Loggers/OwnFilteringChannel.cpp:
```diff
@@ -0,0 +1,51 @@
+#include "OwnFilteringChannel.h"
```
This seems like not finished yet.

petern48 replied:
I just removed the commented-out debug statements. Was that what you were referring to?

antaljanosbenjamin replied:
Ah yes, I will have another look then today.

petern48 committed:
Add getProperty() to OwnFilteringChannel.h and clean up debug comments in .cpp

petern48 commented:
@antaljanosbenjamin Where can I learn what loggers are available (i.e. their names) and what they output? I searched the website and couldn't find it. Below is an example from the [docs](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#logger:~:text=%3C/logger%3E-,Per%2Dlevel%20Overrides,-The%20log%20level) that specify a "Backup" and "RBAC" logger. I'm not sure where these names come from, so I haven't been able to confidently test.
```
<logger>
    <levels>
        <logger>
            <name>Backup</name>
            <level>none</level>
        </logger>
        <logger>
            <name>RBAC</name>
            <level>none</level>
        </logger>
    </levels>
</logger>
```

antaljanosbenjamin commented:
Terrible sorry, I missed your question. The logger names are strictly specified, because
- new loggers can be added and old loggers can be removed on a daily basis
- some logger names are dynamical, they coming from table/part names

The easiest way to find a few loggers is to search in the code for the usage of `getLogger` or just check the logs in on  of the stateless [tests](https://s3.amazonaws.com/clickhouse-test-reports/69657/2360dc00538b97229fbf3829fbe48ad42b38daf2/stateless_tests__tsan__[2_4].html).

antaljanosbenjamin commented on src/Daemon/BaseDaemon.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "Loggers/OwnFilteringChannel.h"
```
This should go into the list of includes not at the first line.

petern48 replied:
Whoops yeah, that must have been auto-generated by my IDE or something. Just fixed.

antaljanosbenjamin commented on src/Loggers/Loggers.cpp:
```diff
@@ -12,6 +13,7 @@
 #include <Poco/Net/RemoteSyslogChannel.h>
 #include <Poco/SyslogChannel.h>
 #include <Poco/Util/AbstractConfiguration.h>
+#include "Common/Exception.h"
```
```suggestion
#include <Common/Exception.h>
```

antaljanosbenjamin commented on src/Loggers/Loggers.cpp:
```diff
@@ -1,5 +1,6 @@
 #include "Loggers.h"
 
+#include "Loggers/OwnFilteringChannel.h"
 #include "OwnFormattingChannel.h"
 #include "OwnPatternFormatter.h"
 #include "OwnSplitChannel.h"
```
```suggestion
#include <Loggers/OwnFilteringChannel.h>
#include <Loggers/OwnFormattingChannel.h>
#include <Loggers/OwnPatternFormatter.h>
#include <Loggers/OwnSplitChannel.h>
```

antaljanosbenjamin commented on src/Loggers/Loggers.cpp:
```diff
@@ -251,6 +267,15 @@ void Loggers::buildLoggers(Poco::Util::AbstractConfiguration & config, Poco::Log
+                    if (auto * regexp_channel = dynamic_cast<DB::OwnFilteringChannel*>(logger.root().get(name).getChannel()))
+                        regexp_channel->setRegexpPatterns(pos_pattern, neg_pattern);
+                    else
+                        throw DB::Exception(DB::ErrorCodes::TYPE_MISMATCH, "Couldn't convert to OwnFilteringChannel.");
```
`TYPE_MISMATCH` is an error code for type mismatch in the data, not in C++ types. I guess this shouldn't happen unless something is wrong, so feel free to use `LOGICAL_ERROR`.

antaljanosbenjamin commented on src/Loggers/Loggers.cpp:
```diff
@@ -347,16 +372,29 @@ void Loggers::updateLevels(Poco::Util::AbstractConfiguration & config, Poco::Log
+    if (auto * regexp_channel = dynamic_cast<DB::OwnFilteringChannel*>(logger.getChannel()))
+        regexp_channel->setRegexpPatterns(global_pos_pattern, global_neg_pattern);
+    else
+        throw DB::Exception(DB::ErrorCodes::TYPE_MISMATCH, "Couldn't convert to OwnFilteringChannel.");
```
Same as above.

antaljanosbenjamin commented on src/Loggers/OwnFilteringChannel.cpp:
```diff
@@ -0,0 +1,47 @@
+    // Apply formatting to the text
+    if (pFormatter)
+    {
+        pFormatter->formatExtended(ExtendedLogMessage::getFrom(msg), formatted_text);
```
I am afraid this can cause some serious issues as we can log a lot of messages. If we format the log messages twice (once here, once before we outputting them), then the logging might be slowed down significantly. Obviously this is only speculation and kind of a worst-case scenario, but it would be nice to see the impact of this in some real examples.

petern48 replied:
That's what I was concerned about. The coloring and formatting are done in one function. I'm not aware of any way we can decouple them into separate functions. Since we don't want two passes, I'm not seeing a perfect solution. I think the next best alternative would be one of the following.

- Skip the filtering process if no patterns are specified. This would isolate the slowdown to only cases where this feature is used. We could add a note about the slowdown in the docs.
(I just implemented this way in my most recent commit, but tests are still failing)

- Only one pass in all cases, but incorrect behavior for colored output and less clean code - I could instead add duplicate filtering channels after each formatter that already exists in the split channel. The color characters would also be passed to the filter since we want the formatted text, misleading the user. We could put a note in the docs to warn about this. (Or maybe we could try filtering the color characters out in that case, but that'll be pretty slow again)

I think at minimum, we should do the first bullet, but I'm not sure about the second one. @antaljanosbenjamin What do you think? Would this be good enough?

antaljanosbenjamin replied:
The first one is a must have as you said and implemented, nicely done. I think the second one is very very questionable. Let's go with this as is right now and see how it turns out. Maybe add a note about this in the docs.

antaljanosbenjamin commented on src/Loggers/OwnFilteringChannel.h:
```diff
@@ -0,0 +1,68 @@
+#include <Poco/Channel.h>
+#include <Poco/Message.h>
+#include <Poco/Util/AbstractConfiguration.h>
+#include "OwnPatternFormatter.h"
```
```suggestion
#include <Loggers/OwnPatternFormatter.h>
```

antaljanosbenjamin commented on src/Loggers/OwnFilteringChannel.cpp:
```diff
@@ -0,0 +1,47 @@
+#include "OwnFilteringChannel.h"
```
```suggestion
#include <Loggers/OwnFilteringChannel.h>
```

antaljanosbenjamin commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -1568,6 +1570,28 @@ The log level of individual log names can be overridden. For example, to mute al
+    <message_regexp>.*executeQuery.*</message_regexp>
+    <message_regexp>.*ConfigReloader.*</message_regexp>
+
+    <levels>
+        <logger>
+            <name>RBAC</name>
+            <!-- For logger 'RBAC', instead of matching for '.*executeQuery.*' and '.*ConfigReloader.*' match instead for '.*Application.*' and '.*Setting.*'. -->
+            <message_regexp>.*Application.*</message_regexp>
+            <message_regexp_negative>.*Setting.*</message_regexp_negative>
+        </logger>
+    </levels>
```
Also, this shouldn't be inside the `level` tag. Logging level is `trace`, `info`, `debug`, `error`, etc. This is not connected to levels.

What do you say about

```
<logger>
    <!-- Global: Only log messages that have 'executeQuery' in them and not 'ConfigReloader' -->
    <message_regexp>.*executeQuery.*</message_regexp>
    <message_regexp>.*ConfigReloader.*</message_regexp>

    <message_regexps>
        <logger>
            <name>RBAC</name>
            <!-- For logger 'RBAC', instead of matching for '.*executeQuery.*' and '.*ConfigReloader.*' match instead for '.*Application.*' and '.*Setting.*'. -->
            <message_regexp>.*Application.*</message_regexp>
            <message_regexp_negative>.*Setting.*</message_regexp_negative>
        </logger>
    </message_regexps>
</logger>
```

petern48 replied:
Sure, just changed the behavior to this.

petern48 committed:
Fix include statement formatting

Co-authored-by: János Benjamin Antal <antaljanosbenjamin@users.noreply.github.com>

petern48 committed:
Fix error codes and include format for code review

petern48 committed:
Merge branch 'ClickHouse:master' into regexp_logger

petern48 committed:
Skip extra formatter if regexp_patterns are empty

petern48 commented:
I made it so that the filtering (and extra formatting pass) is skipped when no regexp pattern is provided, but there's still a decent number of tests that fail. Any ideas? Many just say "Server died." I thought those were due to the slowdown before, but it shouldn't be slow anymore.

Also is there any easy way to run the integration tests locally? I read online that they're set up to run in CICD, so setting it up locally is not straightforward.

antaljanosbenjamin commented:
> Any ideas? Many just say "Server died.

Usually you can see the server logs in our CI. For integration tests download the ZST files from the most right column:
![image](https://github.com/user-attachments/assets/75a5f555-eece-4450-a4b0-ff145b2e6e2c)

For functional (stateless) tests we have all kinds of logs, most importantly `clickhouse-server.err.log` and `clickhouse-server.log.zst` files above the table:
![image](https://github.com/user-attachments/assets/330b30e3-b12c-48a7-8d53-062123cff742)

For this particular case only tests with TSan failed, thus the issue is probably a datarace somewhere. You can find the TSan issue in [`stderr.log`](https://s3.amazonaws.com/clickhouse-test-reports/69657/db1609c48da01356432dcccbf1121e3371e2556a/stateless_tests__tsan__[3_4]/stderr.log) file for a stateless test: `WARNING: ThreadSanitizer: data race (pid=615)`. For the complete message see the linked file or [this pastila](https://pastila.nl/?0001cbb0/be6f85e8756499a4e14d1b6b26c8949a#5BzYHm/XSF1iwt35bCM6sw==).

To run integration tests locally you can use the [`runner`](https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/runner) script, e.g. to run some kafka tests: `./runner --binary /<path-to-your-checkout>/build/programs/clickhouse test_storage_kerberized_kafka/test.py::test_kafka_json_as_string`.

For running the stateless tests you need to start clickhouse and then you can use [`clickhouse-test`](https://github.com/ClickHouse/ClickHouse/blob/master/tests/clickhouse-test), e.g.: `clickhouse-test 03224_invalid_alter`.

petern48 committed:
Try reverting changes to LocalServer.cpp and BaseDaemon.cpp

petern48 committed:
Add mutexes for writing in OwnFilteringChannel.h

petern48 committed:
Delete comments

petern48 committed:
Change config to separate message_regexps tag instead of level tag

petern48 committed:
Add note about slowdown to the docs

petern48 committed:
Add test_regexp_logger integration test and update example in docs

petern48 committed:
Switch configs and update name of test function in test_regexp_logger

petern48 committed:
Fix bug by adding period for logger specific filtering

petern48 committed:
Clean up test_regexp_logger

petern48 committed:
Update docs with note

petern48 committed:
Add const qualifiers and create local var in Loggers.cpp

petern48 committed:
Fix config syntax inside buildLoggers() and cleanup

petern48 committed:
Fix bug with different loggers sharing the same channel

petern48 committed:
Fix flaky test_regexp_logger

petern48 committed:
Add read and write locks to fix race condition in OwnFilterinChannel

petern48 committed:
Restructure so that OwnFilteringChannel is only created when needed

petern48 committed:
Make Poco::Logger getChannel() and setChannel() thread-safe

antaljanosbenjamin commented on docs/en/operations/server-configuration-parameters/settings.md:
````diff
@@ -1576,6 +1578,28 @@ The log level of individual log names can be overridden. For example, to mute al
+```xml
+    <logger>
+        <level>trace</level>
+        <!-- Global: Don't log and Trace messages -->
````
```suggestion
        <!-- Global: Don't log Trace messages -->
```

petern48 replied:
Actually, I just re-read this. The original version was correct. I just updated it to the following to be completely clear. `Global: Don't log messages that have "Trace" in them`

antaljanosbenjamin reviewed (commented):
The changes looks good, if the tests succeed, I will check out the integration test more thoroughly.  Otherwise everything looks good!

petern48 committed:
Update docs/en/operations/server-configuration-parameters/settings.md

Co-authored-by: János Benjamin Antal <antaljanosbenjamin@users.noreply.github.com>

antaljanosbenjamin commented on tests/integration/test_regexp_logger/test.py:
```diff
@@ -0,0 +1,86 @@
+        node.query("SYSTEM RELOAD CONFIG")
+        node.query("SELECT 1")
+
+    log = get_log(node)
+    assert re.search(r".*Loaded config.*", log)
+    assert re.search(r".*executeQuery.*Read.*", log)
+    assert re.search(r".*executeQuery.*from.*", log)
```
Maybe try `contains_in_log` and `rotate_logs` in the tests.

antaljanosbenjamin commented on tests/integration/test_regexp_logger/test.py:
```diff
@@ -0,0 +1,86 @@
+
+    node.replace_config("/etc/clickhouse-server/config.d/log.xml", updated_config)
+    node.query("SYSTEM RELOAD CONFIG;")
+    node.exec_in_container(
+        ["bash", "-c", "> /var/log/clickhouse-server/clickhouse-server.log"]
+    )
```
This can be achieved with `instance.rotate_logs()`.

antaljanosbenjamin commented on tests/integration/test_regexp_logger/test.py:
```diff
@@ -0,0 +1,86 @@
+    log = get_log(node)
+    assert len(log) > 0
+
+    assert not re.search(r".*Loaded config.*", log)
+    assert re.search(r".*executeQuery.*Read.*", log)
+    assert not re.search(r".*executeQuery.*from.*", log)
```
Same as above: `contains_in_log` and `rotate_logs`.

petern48 committed:
Use rotate_logs() and contains_in_log() in test_regexp_logger

petern48 commented:
@antaljanosbenjamin I implemented those two changes you requested (and only those changes), and now the "Docs check" keeps failing (it also happened one other time for a previous commit I've reverted, which you can't see anymore). The logs mention broken links completely unrelated to the PR. Is this an issue with CICD or is there some niche issue with the tiny doc change we made?

petern48 committed:
Empty commit to trigger CICD

petern48 commented:
@antaljanosbenjamin All passed! Seems like it was just an issue with CICD

petern48 commented:
Just some comments for myself in case I look back later and wonder what was going on, since the task turned out to be much harder than expected.

Problems encountered and solved:
- (Main Problem) Fixed a major hidden bug (details below) that would cause problems for this and any future features that involves the modifying the logger channels
- Fixed the issue of slow performance when this feature is not in use
- Fixed concurrency issues with locking, including in the Poco Logger library
- Fixed memory leaks caused by advanced usage of `Poco::AutoPtr` (shared pointer)
- Fixed flaky tests issues

Details about the "Main Problem" in the first bullet:
- There was a very hidden bug that caused of incorrect behavior when specifying multiple loggers: Most (but not all) loggers were created elsewhere in the codebase (spread out) after the call to `buildLoggers()`, and many shared the same channel. This meant we couldn't rely on `buildLoggers()` to set things up correctly, so I made it so this feature would create a new channel (`OwnFilteringChannel`) specific to that logger, and it only does so for the first time it's needed for that specific logger (if ever). It may happend in `updateLevels()` in addition to `buildLoggers()` to fix the problem of some loggers being created after `buildLoggers()`

Pretty proud of this PR :)
